### PR TITLE
rejectOnEmpty mode

### LINKED ISSUE #272: flag to emit an error if select result is null
hey,

I'm using async.js waterfall method together with sequelize to prevent 'callback chains of doom'. One scenario I'm regularly confronted with is that the first call in a chain is a select query and the following parts of the chain deal with the result of this select ... most of the time this results in code like this:

```javascript
async.waterfall([...
  function(callback) {
    Client.find(:id).done(function(err, res) {
      if(err !== null) callback(err);
      else if(res !== null) callback(new Error('Sorry no DB entry could be found!'));
      else callback(null, res);
    }):
  }, 
  function(client, callback) {...
```

... it would be nice to have an extra flag in Sequelize to produce an error if there is no result available so I could just pass trough my callback to done() and handle the errors in async's final method.

### LINKED ISSUE #5480: Find returns a null instance instead of catch
So this is my section of code:
```JavaScript
  /**
   * Verifies the specified user against the specified token
   * @param {string | number} user
   * @param {string} token
   * @returns {Object} promise
   */
  static verifyEmailVerificationToken(user, token) {
    return models.User.findOne({where: {$or: [{email: user}, {id: user}], emailToken: token}}).then(function(user) {
      return user.update({emailConfirmed: true, emailToken: null}).then(function(user) {
        eventService.emit('users.verifyEmailVerificationToken', user);
        return user;
      });
    });
  }
```

The `user.update` section should not be fired am I correct? The promise's "catch" callback should be the code that is triggered, however it is not. The "then" callback is being called with a null instance.

now I know this is the intended purpose, but shouldn't it invoke the catch instead?

### PR DESCRIPTION
**Pull request by sushantdhiman**
Allow finder methods to throw `RecordNotFound` error when query returns null set and `kenophobic` mode is `true`. This option can be set per method basis or with model options.

Closes #5480 
Closes #272

> Kenophobic means a person with fear of emptiness or void, You can suggest any other name if want to :)

### CODE REVIEW

**Comment by mickhansen** [comment]
While i like the name i think `throwOnEmpty` or `rejectOnEmpty` would be more fitting.
Could you please also extend the API to allow a custom error from the user as an argument? (Would be helpfull for people doing error handling in APIs wanting to pass a 404 erorr instance

File: lib/model.js [hunk 1]
```diff
@@ -1409,6 +1415,9 @@ Model.prototype.findAll = function(options) {
 };
 
 Model.$findSeparate = function(results, options) {
+  if(_.isEmpty(results) && options.kenophobic)
+    throw new sequelizeErrors.RecordNotFoundError();
+
```
Comment (mickhansen) [inline lib/model.js]: Lets place this in the find call. findSeperate deals with other stuff.

**Comment by sushantdhiman** [comment]
Hi @mickhansen , I am handling the `kenophobic` mode from `findAll` now, not using the `$findSeparate` anymore. Also custom error instance can be passed to `kenophobic` flag and it will be thrown, added tests for that. Please review

File: lib/model.js [hunk 2]
```diff
@@ -1404,6 +1410,16 @@ Model.prototype.findAll = function(options) {
       return this.runHooks('afterFind', results, options);
     }
   }).then(function (results) {
+
+    //kenophobic mode
+    if(_.isEmpty(results) && options.kenophobic) {
+      if( (typeof options.kenophobic === 'function')) {
+        throw new options.kenophobic();
```
Comment (mickhansen) [inline lib/model.js]: It could also be an already init'e derror instance: `kenophobic: new Status404Error('No such user')`

File: lib/errors.js [hunk 3]
```diff
@@ -299,3 +299,15 @@ error.InstanceError = function (message) {
...
+/**
+ * Thrown when a record was not found, Usually used with kenophobic mode (see message for details)
+ * @extends BaseError
+ * @constructor
+ */
+error.RecordNotFoundError = function (message) {
+  error.BaseError.apply(this, arguments);
+  this.name = 'SequelizeRecordNotFoundError';
+  this.message = message;
+};
```
Comment (mickhansen) [inline lib/errors.js]: `RecordNotFound` implies single result, maybe a better generic would be `EmptyResultError`

**Comment by mickhansen** [comment]
We probably still need an option name change, /cc @janmeier

File: lib/model.js (hunk 2 above)
Comment (sushantdhiman) [inline lib/model.js]: Not sure if I have done it in right way, but this will work for both cases now. Please check

**Comment by sushantdhiman** [comment]
I am good with of any of these options

* `rejectOnEmpty`
* `throwOnEmpty`
* `kenophobic`

`kenophobic` is bit inspired by `paranoid`, your comments @janmeier

File: lib/model.js (hunk 2 above)
Comment (mickhansen) [inline lib/model.js]: yeah looks good

File: lib/model.js [hunk 4]
```diff
@@ -1404,6 +1410,18 @@ Model.prototype.findAll = function(options) {
       return this.runHooks('afterFind', results, options);
     }
   }).then(function (results) {
+
+    //kenophobic mode
+    if(_.isEmpty(results) && options.kenophobic) {
+      if( (typeof options.kenophobic === 'function')) {
```
Comment (mickhansen) [inline lib/model.js]: space after `if` (goes for entire diff), no space after `(`, no need to have double `()` here actually.

File: test/integration/model/findAll.test.js [hunk 5]
```diff
@@ -1389,4 +1389,42 @@ describe(Support.getTestDialectTeaser('Model'), function() {
...
+      return Model.sync({ force: true })
+        .then(function() {
+          return expect(Model.findAll({
+            where: {
+              username: 'some-username-that-is-not-used-anywhere-for-sure-this-time'
+            }
+          })).to.eventually.be.rejectedWith(Sequelize.ConnectionError);
+        });
+    });
+  });
```
Comment (mickhansen) [inline test/integration/model/findAll.test.js]: We need an extra unit test so that we handle all 3 cases, none passed, class passed or instance passed

**Comment by sushantdhiman** [comment]
@mickhansen I have fixed code styling issues, Added new tests to check cases of `Initialized`, `Instance` and `undefined` value for `kenophobic` flag

**Comment by sushantdhiman** [comment]
@janmeier waiting for you :)

**Comment by mickhansen** [comment]
Will give him another day or so and we'll just make a decision.

File: lib/model.js [hunk 6]
```diff
@@ -1353,7 +1356,10 @@ Model.prototype.findAll = function(options) {
   tableNames[this.getTableName(options)] = true;
   options = optClone(options);
 
-  _.defaults(options, { hooks: true });
+  _.defaults(options, { hooks: true, kenophobic: false });
```
Comment (janmeier) [inline lib/model.js]: Couldn't this be ` _.defaults(options, { hooks: true, kenophobic: this.options.kenophobic });`?

**Comment by janmeier** [comment]
I lol'd at kenophobic, but I think `rejectOnEmpty` will be better ;)

File: lib/model.js (hunk 6 above)
Comment (sushantdhiman) [inline lib/model.js]: Yes it can be, it will save one line of code too

**Comment by sushantdhiman** [comment]
`rejectOnEmpty` :( , Ok I will do it

File: lib/model.js (hunk 6 above)
Comment (sushantdhiman) [inline lib/model.js]: Just tested, I didn't knew `defaults` behavior, but it will only overwrite `undefined` values, even if value is `false` it will not assign the `true` value from model options.

**Comment by mickhansen** [comment]
@sushantdhiman Hey, why'd you close this? This was about ready to merge (just forgot about it since commits don't send me a notification :()

**Comment by sushantdhiman** [comment]
Hi @mickhansen , I know you are busy but please review the pull requests when you get chance, if they are merged on time I feel encouraged to work on new one.

**Comment by mickhansen** [comment]
Of course @sushantdhiman, but remember that commits don't notify me so i can't know you've made changes unless you comment.

If you can simply squash to a few descriptive commits (or change the descriptions to reflect the new name) we should be ready to go.

**Comment by sushantdhiman** [comment]
Hi @mickhansen , I have changed the commit messages and squashed them

**Comment by mickhansen** [comment]
Thanks @sushantdhiman, great work.

**Comment by sushantdhiman** [comment]
Thanks @mickhansen
